File: #1061.lexicographically-smallest-equivalent-string.cpp
class Solution {
public:
    string smallestEquivalentString(string s1, string s2, string baseStr) {
        
        int d[26]{};
        for (int i{}; i < 26; ++i) {
            d[i] = i;
        }
    
        for (int i{}; i < size(s1); ++i) {
            int x{s1[i] - 'a'};
            for( ; x != d[x]; x = d[x]);

            int y{s2[i] - 'a'};
            for( ; y != d[y]; y = d[y]);

            d[std::max(x, y)] = std::min(x, y);
        }

        for (auto &c : baseStr) {
            int x{c - 'a'};
            for ( ; x != d[x]; x=d[x]);
            c = 'a' + x;
        }           
        return baseStr;
    }
};

File: #1061.lexicographically-smallest-equivalent-string.js
/**
 * @param {string} s1
 * @param {string} s2
 * @param {string} baseStr
 * @return {string}
 */
var smallestEquivalentString = function(s1, s2, baseStr) {
    const union = new Map();
    const unionFind = (target) => {
        const value = union.get(target) ?? target;
        return value === target ? value : unionFind(value);
    };

    for (let index = 0; index < s1.length; index++) {
        const a = unionFind(s1[index]);
        const b = unionFind(s2[index]);

        a < b ? union.set(b, a) : union.set(a, b);
    }

    return [...baseStr].reduce((result, str) => {
        return result + unionFind(str);
    }, '');
};
